Jwt_authn config update: Support per route Jwt requirements

qiwzhang opened the pull request:
Signed-off-by: Wayne Zhang <qiwzhang@google.com>

*Description*:
Based on the requirement discussion from https://github.com/envoyproxy/envoy/issues/2514.
Change the Jwt_authn config to support different requirement based on route match.

*Risk Level*: Low

qiwzhang commented:
@llchan please help to review this config change to see if it works for your cases.
@diemtvu, @lizan  Please help to review this change.

lizan commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,33 +183,67 @@ message JwtHeader {
+  // this field will not be used.
+  // If true, the request is allowed if JWT is missing or JWT verification fails.
+  // Default is false, a request without JWT or failed JWT verification is not allowed.
+  bool allow_missing_or_failed = 2;
```
Just delete this instead of `DEPRECATED` since no one is using this? same below.

qiwzhang replied:
Unfortunately, Istio 0.8 is using this one.  I am pretty sure nobody is using  the "bypass" field.

diemtvu replied:
We actually do
https://github.com/istio/istio/blob/0214aee9cb9c75afa59e83364c7acd52afa01656/pilot/pkg/networking/plugin/authn/authentication.go#L128

diemtvu commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,33 +183,67 @@ message JwtHeader {
+// PerRouteConfig for jwt_authn filter uses this message.
+message JwtRequirement {
+  // Specify list of required Jwt issuers. If list is empty,  Jwt is NOT required.
+  repeated string required_issuers = 1;
```
Can you add comment to clarify if all are required or just one.

qiwzhang replied:
All are required.  Updated the comment.

diemtvu commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,84 @@ message JwtHeader {
+  // It only applies when required_issuers field is not empty.
+  // If true, the request is allowed if JWT is missing or JWT verification fails.
+  // Default is false, a request without JWT or failed JWT verification is not allowed.
+  bool allow_missing_or_failed = 2;
```
It's not quite clear to me why we need this bool? Can we simply put empty required_issuers list?

qiwzhang replied:
Sound likes a good idea.  I will remove it.

llchan commented:
@qiwzhang This works for me, but will you also include the ability to configure at the route level with per-route-filters as discussed on the issue thread? I can see other people wanting that. If we support both it should cover a pretty wide variety of setups.

llchan commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,82 @@ message JwtHeader {
 // [#not-implemented-hide:]
 message JwtAuthentication {
   // List of JWT rules to valide.
   repeated JwtRule rules = 1;
```
I would maybe rename this, since they are no longer rules but provider defintions.

llchan replied:
Actually I don't know anymore, maybe rules is okay, you guys can discuss internally what the naming should be.

qiwzhang replied:
This config is used by Istio now.   Field change requires code change.

qiwzhang replied:
Changed to Provider

llchan commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,82 @@ message JwtHeader {
+  // applied. If there are overlapped match conditions, please put the most specific match first.
+  // If PerRouteConfig is specified for the route, this field will not be uesd.
   //
-  // Examples: bypass all CORS options requests
```
What's the plan with other kinds of request matching (e.g. header matching for CORS options requests)?

qiwzhang replied:
envoy.api.v2.route.RouteMatch supports header matching in additional to path matching.

llchan commented:
Not sure how to comment on parts of the file that aren't part of the diff so I'll do it here.

I think the issuer/audience definitions (aka `JwtRule` currently) need an additional user-specified key/id that is used when specifying the requirements. The issuer url/email alone may not be flexible enough because the audience is also important for validating a token. For example, if I have one issuer https://issuer.com that issues tokens for two audiences foo.com and bar.com, those are two different specs and the user should be able to refer to them independently in the rules.

qiwzhang commented:
I added "audiences" in JwtRequirement to override the one in the JwtProvider.  You can set different audiences in the per-route requirement.

qiwzhang commented:
@llchan, @lizan, @htuch  PTAL.   I like to get this in.  Thanks

llchan commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,126 @@ message JwtHeader {
+//      - requires:
+//          issuer: jwt_issuer2
+//          audiences: [audience1, audience2]
+//
```
I think the YAML examples above are a bit incorrect ("requires" field, dashes should be on issuer lines I think), could you double check @qiwzhang?

llchan commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,126 @@ message JwtHeader {
+//        - requires:
+//            issuer: issuer2
+//
+//// [#not-implemented-hide:]
```
Same as above, check the YAML, I think "requires" fields are a little off.

qiwzhang replied:
Done

llchan commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,126 @@ message JwtHeader {
+  //      match:
+  //         prefix: "/"
+  //      - requires:
+  //         issuer: issuer1
```
Again, check the YAML, I think both match and requires need some tweaks.

qiwzhang replied:
Done

llchan commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,151 @@ message JwtHeader {
+  // be rejected.
+  // If a Jwt is not required, but it is in the providers list, it will be verified.
+  // But it will not fail the request if its verification fails.
+  repeated JwtRequirement requires = 2;
```
If I'm reading the comment correctly, this means we're AND-ing all the requirements? Is that more useful than OR-ing the requirements? I'd think that a typical request has one JWT and would thus only match one issuer, so it's unlikely for AND-ed requirements to be satisfied unless they all have the same issuer and only differ in audience + the JWT has multiple audiences defined.  Having OR-ed requirements means the route can accept tokens from multiple issuers, which seems like a more common use case. Thoughts?

qiwzhang replied:
I think AND-ing the requirement will be more useful.   If you ever want to verify two JWT from different issuers,  you most likely want both of them to pass, not just one of them.   For example,  Google cloud supports IAP and end-user token,  some requests want both to pass, it will be: [iap,  end_user_jwt],  some only want iap,  it will be [iap],   even the requests may carry end_user_jwt,  but it is not required.  It is very hard for me to imaging a request only wants either iap or end_user_jwt.

llchan replied:
I see, so the request carries multiple JWTs in different headers/params? If so AND-ing makes sense, because otherwise there isn't a good way to specify it.

lizan replied:
@qiwzhang we do OR-ed requirements in Google Cloud, https://github.com/googleapis/googleapis/blob/master/google/api/auth.proto#L61 . It is useful to support AND-ing when you have multiple JWTs. However, more common cases we've seen are: your service serve both internal service-to-service traffic and ingress traffic, then you need to specify `JWT signed by a service account` OR `End-user JWT issued by A` OR `End-user JWT issued by B`.

llchan replied:
Agree with @lizan , I'd prefer OR too, at least at the top level. Just seems more common to allow multiple issuers to authenticate a user.

However, I could see multi-JWT requests being used, so perhaps we can come up with something that satisfies both? I'd rather start off with a slightly more verbose config vs change it down the road.

A few misc notes that come to mind for multi-JWT configuration:
- We could do a "2-d" list, where the top level iterates over OR-ed requirements and the inner is AND-ed. Or maybe use a union or oneof?
- We'd probably want a way to map headers to requirements explicitly, otherwise the implementation would have to try each of the N^2 pairs. So either the requirements need to specify the `from_headers` or `from_params`, or the requirements need ids and there needs to be some mapping elsewhere.

lizan replied:
Let's not make things too complicate, we might be fine with OR here, but if there are multiple JWTs, validate all of them, and let authZ (RBAC) to handle the complex AND/OR rules.

llchan commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,151 @@ message JwtHeader {
+  // this field will not be used.
   // If true, the request is allowed if JWT is missing or JWT verification fails.
   // Default is false, a request without JWT or failed JWT verification is not allowed.
   bool allow_missing_or_failed = 2;
```
I vaguely remember someone else commenting on this field and recommending that it be removed (rather than deprecated) since this config is still in the alpha stage. Was there a final decision on that? It seems that this behavior is pretty trivial to specify as a single catch-all requirement.

qiwzhang replied:
Istio is using it (as internal config between its components, not a user facing config).  So we could not remove it now,  but like to remove it once Istio code is changed to use new "rules" field.

llchan replied:
Got it, sounds good.

qiwzhang commented:
@llchan could you approve this PR?  
@lizan @htuch could you PTAL?

qiwzhang commented:
OK,  I added 2D  "or_jwt" and "and_jwt" in the JwtRequiement.    PATL

lizan commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,178 @@ message JwtHeader {
+  // Specify list of required JwtProviders. Their results are AND-ed;
+  // all of them must pass, if one of them fails or missing, it fails.
+  message AndRequirement {
+    repeated RequiredProvider and_jwt = 1;
```
add validate rule that this must not be empty.

I feel there should be better name than `and_jwt` / `or_jwt`...

qiwzhang replied:
Added validate rule, and renamed to "one_of" and "all_of"

qiwzhang commented:
Format failure is from someone else files.

ERROR: clang-format check failed for file: ./api/envoy/config/grpc_credentials/v2alpha/file_based_metadata.proto

lizan commented:
@qiwzhang https://github.com/envoyproxy/envoy/pull/3517 should fix the format ci, which is broken on master now.

llchan commented:
Overall I think this is reasonable. It feels a bit verbose and seem to be overreaching a bit into authz, but at least it's pretty clear what it's doing.

That said, I'm not 100% sold on it, so I'll suggest another option that isn't quite as flexible, but is far more concise for the common use case while supporting simple AND-ed requirements:
```
- rules:
  - match: { prefix: "/foo" }
    requires_any: [{ issuer: "iss-A" }, { issuer: "iss-B" }]
  - match: { prefix: "/bar" }
    requires_all: [{ issuer: "iss-A" }, { issuer: "iss-B" }]
```
Compare that to the equivalent one_of/all_of config:
```
- rules:
  - match: { prefix: "/foo" }
    requires: { one_of: [{ all_of: [{ issuer: "iss-A" }] }, { all_of: [{ issuer: "iss-B" }] }] }
  - match: { prefix: "/bar" }
    requires: { one_of: [{ all_of: [{ issuer: "iss-A" }, { issuer: "iss-B" }] }] }
```
This should be doable with a simple `oneof` between `requires_any` and `requires_all`. Or `requires_one` if that's better.

FWIW I'm also in favor of what @lizan suggested: drop AND-ed requirement support at this layer and just use a simple OR requirement, but validate all JWTs presented. That would certainly suffice for me, but since I'm not familiar with the multi-JWT use case I'll defer that decision to you guys.

qiwzhang commented:
@llchan  I like your idea.  I kept "requires_any" and "requiers_all" to support multiple JWT requirements. 
Unfortunately,   proto oneof type doesn't allow repeated field.   So "requires_any" and "requiers_all" has to be in this format:  
```
requires_any: { list: [issuer: "issuer1"], [issuer: "issuer2"] }
```
It feels weird, but they are rarely used. It should be OK.
PTAL.  If it is OK, please approve it.

htuch commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,162 @@ message JwtHeader {
+
+    // Specify list of RequiredProvider. Their results are AND-ed.
+    // All of them must pass, if one of them fails or missing, it fails.
+    RequiredProviderList requires_all = 4;
```
Have you considered a tree of protos with AND/OR semantics? Similar to what we do for log filtering, see https://github.com/envoyproxy/envoy/blob/c15019e79c832d9f0a09468affaadabc4be3e115/api/envoy/config/filter/accesslog/v2/accesslog.proto#L136

llchan replied:
This seems like it could fit the bill. The only real change from where the PR is now is an additional oneof message wrapper, right?

```
- rules:

  # require A
  - match: { prefix: "/baz" }
    requires: { issuer: "iss-A" }

  # require A or B
  - match: { prefix: "/foo" }
    requires_any:
      requirements:
      - requires: { issuer: "iss-A" }
      - requires: { issuer: "iss-B" }
  
  # require A and B
  - match: { prefix: "/bar" }
    requires_all:
      requirements:
      - requires: { issuer: "iss-A" }
      - requires: { issuer: "iss-B" }

  # require A and (B or C)
  - match: { prefix: "/qux" }
    requires_all:
      requirements:
      - requires: { issuer: "iss-A" }
      - requires_any:
          requirements:
          - requires: { issuer: "iss-B" }
          - requires: { issuer: "iss-C" }
```

I have a slight preference to inline the oneof into both the `RequirementRule` and the new message type, that way we can avoid the extraneous nested level, but I suppose that's not very DRY so I can understand either way.

qiwzhang replied:
Like the idea.  Update the config file to support it.
I

htuch commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,162 @@ message JwtHeader {
-  // If true, the request is allowed if JWT is missing or JWT verification fails.
-  // Default is false, a request without JWT or failed JWT verification is not allowed.
-  bool allow_missing_or_failed = 2;
+  // This audiences field overrides the one in JwtProvider if not empty.
```
What is meant here by overriding? E.g. does this make the match more specialized than if just using the JwtProvider?

qiwzhang replied:
Semantic wise, audiences specified in the requirement will override the ones specified in the Provider.   
@llchan do you know how such audience override can be used?

htuch commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,162 @@ message JwtHeader {
-  repeated JwtRule rules = 1;
+// This message specifies a Jwt requirement.
+message RequiredProvider {
+  // Specify the issuer for the required JwtProvider.
```
This uniquely identifier a single JwtProvider already specified elsewhere?

llchan replied:
I don't believe the current proposal has a unique identifier, it's using the issuer as the id and allowing overrides on audiences.

If we model this closer to the googleapis model, we'd add a user-specified id which allows for multiple provider configurations with the same issuer (e.g. with different audience sets). This also addresses the other comment about audience overrides. I'm in favor of this, but don't feel very strongly about it.

qiwzhang replied:
Yes, "issuer" uniquely identifies the Provider specified in the "providers" field. 

Hi @llchan based on my understanding of googleapis AuthRequirement,  it is similar to what I have proposed here. Here is its example.

-- selector:  /foo
   requirements:
      provider_id: issuer1
      audiences: [audience1, audience2]

llchan replied:
The google api I'm referring to is [this one](https://github.com/googleapis/googleapis/blob/master/google/api/auth.proto#L79). Notice that the "primary key" is not the issuer, but rather a user-defined id. This allows multiple configurations to be defined per issuer, so the routes can refer to those specific configurations rather than overriding a single master config for that issuer.

I think it's fairly unlikely anyone needs really really fine grained per-route audience overrides. More likely, there's some big giant issuer that can issue tokens for multiple audiences, and routes require a particular issuer+audience pair. By letting the user id-ify these pairs we allow route requirements to refer to shared configurations.

llchan replied:
Also in case it wasnt clear, id-ifying these obviates the need for the audience overrides, which I think makes the whole thing a little more clear.

lizan commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -30,7 +31,30 @@ import "validate/validate.proto";
+  //      - inline_string: "ACADADADADA"
+  //
+  // Use 9 as the proto index number to support backward compatibility.
+  string id = 9 [(validate.rules).string.min_bytes = 1];
```
perhaps `name` is more consistent across envoy configs?

qiwzhang replied:
Done

lizan commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +207,211 @@ message JwtHeader {
+message JwtRequirement {
+  oneof requires_type {
+    // Specify a required provider id.
+    string provider_id = 1;
```
Make this as a message and including `provider_name`, `audiences`. You are likely to have same issuer with different audiences at different route. So moving audience to this level will make the provider config more reusable.

qiwzhang replied:
As you are aware that newly added provider_name field allows users to create multiple providers with different audiences for the same issuer.  I feel adding audiences in the JwtRequirment is adding another layer of override, it is introducing too much complexity in term of implementation and easy to use.  Let us not do that for now,  we can add it later if it is really necessary.
@llchan what do you think?

lizan replied:
Yes I am aware of that you can create multiple provider for same issuer. Moving the audience from each provider to here, so you don't have to do that for each route that requires just different audience from same issuer.

qiwzhang replied:
If we move audiences from Provider to ProviderRequirement,  it means removing the one in Provider,  it will have problem for Istio Authn.   Based on your idea of "empty rules" means "allow_missing_and_fail = true" for authn.    It will not have any rules,  so it will not have ProviderRequirements defined,   you could not specify "audience" to verify.

lizan replied:
Let me clarify, I meant in whole, sth like this:

```proto

message JwtRule { // better name wanted...
  string provider_name;
  repeated string audiences;
}

message NoRequirement {
}

message JwtRequirement {
  oneof require_type {
    NoRequirement no_require;
    JwtRule rule;
    JwtRequirementOrList requires_any;
    JwtRequirementAndList requires_all;
  }
}
```

qiwzhang replied:
I understood your idea.  If we MOVE audiences to JwtRule,  for Istio authn usage, it will not define any JwtRequirement, so it will not define any JwtRule.  If audiences are only in JwtRule, they will not be verified.

lizan commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +207,211 @@ message JwtHeader {
+  // "one_of" requirements, if the first requirement passes, the rest requiements will not be
+  // verified. But if this flag is true, all requirements will be verified even all of them
+  // passed.
+  bool allow_missing_or_failed = 2;
```
Can we nuke this flag by adding an empty requirement in `RequrirementRule` so `provider: issuser-A` OR  `empty` means same as this? Sounds like it gives more flexibility.

qiwzhang replied:
Good idea.  will update it.

llchan commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +207,209 @@ message JwtHeader {
+  //
+  repeated JwtProvider providers = 1;
+
+  reserved 2;
```
Forgive me if this is obvious, but why is this here? Is there something already in mind for this id, and is there a reason it shouldnt go at the end?

llchan replied:
Is binary compatibility necessary? Seems that these protos are substantially different from what was in istio, and they shouldn't be expected to interoperate. Maybe there's a lot going on over there that I haven't seen though.

lizan replied:
We can clean up those when this proto become v2 or v2beta, keeping this for now if istio can manage the compatibility.

qiwzhang replied:
I think we don't need to support binary compatible.  Filter name will be different.  Istio Pilot need to change filter name anyway.

qiwzhang commented:
Last SHA (83b4aa3) made breaking change. It changed "repeated providers" to map<string, provider>.  It will not be backward compatible with older version Istio is using.
 Action items for Istio is: to move older version of this config.proto back to Istio/api repo and made Istio/proxy and istio/istio use that version.  This will allow Envoy Jwt_authn config to make breaking change.

nickrmc83 commented:
Yes. No problem. I'll take a look tomorrow.

On Thu, 14 Jun 2018, 19:26 Wayne Zhang, <notifications@github.com> wrote:

> @nickrmc83 <https://github.com/nickrmc83>, could you help to review this
> jwt_authn file config change to see if it fixes with your use cases?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/envoyproxy/envoy/pull/3381#issuecomment-397393063>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABhXUBFwPCGKT-X6AqZz8jJ5VBTvIhghks5t8qrVgaJpZM4T-pE->
> .
>

nickrmc83 commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -30,7 +30,7 @@ import "validate/validate.proto";
 //
 // [#not-implemented-hide:]
-message JwtRule {
+message JwtProvider {
```
@qiwzhang From a configuration PoV the only part of this change that doesn't exactly meet my usecase is that the JwtProvider type describes both the verification and forwarding parameters for a JWT. As a result filters that wish to re-use just the verification logic are also required to set the forward rule to true (as false is the default) so that the headers are not mutated. It'd be nice if either forwarding JWT headers was the default behaviour or alternatively and preferentially there was a clear type separation between verification configuration and forwarding configuration. The latter case would allow other filters to expose just the verification configuration as part of their own config. For example:

```
message VerificationRules {
  string issuer = 1 [(validate.rules).string.min_bytes = 1];
  ...
}

message FrowardingRules {
  bool forward = 1;
  string forward_payload_header = 2;
  ...
}

message JwtProvider {
  VerificationRules verification_rules = 1;
  ForwardingRules forwarding_rules = 2;
}
````

nickrmc83 replied:
> Maybe you will not use jwt_authn filter as a whole filter, you will just link its code as c++ library 
 into your filter? I don't think it is a good way to go.

Unfortunately I think the only approach for my filter is to include and link the jwt_authn code directly and invoke it in the same way as http/jwt_authn/http_filter.cc:decode_headers. The reason for this is because the JWT on which I need verification performed is obtained by my filter via a filter-initiated external asynchronous HTTP request. As a result the JWT that is acquired does not pass through the filter chain. So using the code as a filter is not really an option. As such I need to make my filter expose the necessary config to perform verification and ideally not expose parameters that are not relevant whilst not redefining similar structures that already exist. In my case the ```forward``` field is not applicable. A solution to that is simply to document that setting this field within my filter will have no effect. And that's fine. So disregard my initial request but I will be wanting to directly invoke the Authenticator code.

nickrmc83 commented on source/extensions/filters/http/jwt_authn/authenticator.cc:
```diff
@@ -233,22 +234,14 @@ void AuthenticatorImpl::verifyKey() {
-  } else {
-    callback_->onComplete(status);
-  }
+  callback_->onComplete(status);
```
For use by other filters, is it possible to have an onComplete callback that passes a const reference/pointer to the token as well as the status? In my uses case (OpenID Connect) I need to check some additional JWT claims to validate against XSRF attacks. These claims are potentially use case specific so pushing them into the Authenticator would be wrong.

qiwzhang replied:
Clearly, you were thinking about linking jwt_authn code into your filter. 

I will prefer to explore if you can use jwt_authn code as a filter.  Your filter can use its verification result.   Istio_authn originally linked jwt_authn code into its filter,  it makes the filter very complicated since it needs to deal with aync callbacks.  We abandoned that idea.  Now it will not touch jwt_authn filter, just config it to behave the way istio_authn wanted and uses its result.

Now the result is passed by HTTP headers.  We can change it to use requestinfo.dynamidMetadata to pass the result. https://github.com/envoyproxy/envoy/blob/master/include/envoy/request_info/request_info.h#L282

lizan replied:
In (near) future we will make the verified JWT into metadata, that will allow JWT works with RBAC filter, see #3638 too. I think that will address your use case? @nickrmc83

nickrmc83 replied:
@qiwzhang @lizan see my earlier comment about why I can't really re-use the code as a stand-alone filter and why I would like to directly invoke the Authenticator verification code. I suggest pushing forward this PR and discussing my usage in a PR I raise as I think my needs will become clearer.

qiwzhang commented:
@nickrmc83 . It seems that we agreed that your user case is so special, we should not adjust jwt_authn config for that. 
You may raise a separate PR for the Authenticator class code change to separate verification with token forwarding

qiwzhang commented:
@lizan I addressed your comment of moving "audiences" to RequiredProviider by adding a new field in JwtRequirement onof field:  "provider_with_audiences".   Users can config audiences in JwtProvider, they can also override them in JwtRequirement.

qiwzhang commented:
@lizan @htuch  This PR is ready for review.  This PR will just get the config change in first. Its added "JwtRequirement" feature is not implemented yet.  I will use a separate PR to implement it.

htuch commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,219 @@ message JwtHeader {
+
+// This message specifies a list of RequiredProvider.
+// Their results are OR-ed; if any one of them passes, the result is passed
+message JwtRequirementOrList {
```
Out of curiosity, would this work if we just inline it into the above fields? E.g. if we had `repeasted JwtRequirement requires_any = 3 [constraints];`? I'm not familiar with the proto rules on self-reference.

qiwzhang replied:
Not, it will not work,  proto "oneof" doesn't "repeated" field.

htuch commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,219 @@ message JwtHeader {
+//         prefix: "/healthz"
+//
+// In above example, "requires" field is empty for /healthz prefix match,
+// it means that requrests matched the path prefix don't require jwt authentication.
```
Nit: I think what you want is "that requests matching the path prefix don't require JWT authentication."

qiwzhang replied:
Done

htuch commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,219 @@ message JwtHeader {
+  // will pass even the verifications fail. A typical usage is: this filter is used to only verify
+  // JWTs and pass the verified JWT payloads to another filter, the other filter will make decision.
+  // This usage has impact on the process flow. For example, normally for "requires_any"
+  // requirements, if the first requirement passes, the rest requiements will not be verified. But
```
Nit: "the rest of the requirements"

qiwzhang replied:
Removed

htuch commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,219 @@ message JwtHeader {
+  //           - provider_name: "provider2"
+  //
+  // A special handling if rules is empty: all JWTs in a request will be verified, but the request
+  // will pass even the verifications fail. A typical usage is: this filter is used to only verify
```
Nit: "even if the verification fails"

qiwzhang replied:
Removed

htuch commented on api/envoy/config/filter/http/jwt_authn/v2alpha/config.proto:
```diff
@@ -183,44 +183,219 @@ message JwtHeader {
+  //           - provider_name: "provider2"
+  //
+  // A special handling if rules is empty: all JWTs in a request will be verified, but the request
+  // will pass even the verifications fail. A typical usage is: this filter is used to only verify
```
I think this is pretty surprising behavior; is there no way to design this based on principle-of-least-surprise? E.g. add an explicit "continue_request_on_verification_fail" field?

qiwzhang replied:
Thanks for the input. Agreed.  I removed it and add "allow_missing_or_failed" as one of "oneof" in the JwtRequirement.

atali commented:
Hello guys, 

I would like to support multi-tenancy on our microservice. Once the user is authenticated, my JWT will contains a field  like tenant_id: 12323232 . I would like to route the request base on that jwt field. Do you think it will work on what you have done ?

lizan commented:
@atali can you open an issue with more details of your use case? Currently it is not possible to route request based on JWT claims (this PR is reverse, route-based JWT configuration).

potatop commented:
Is this still being worked on? If not, I would like to help.

qiwzhang commented:
@potatop could you be specific on which feature you like to help?
for routing based on jwt fileds,  I am not aware of anybody is working on it. 
for supporting per-route requirement: config proto is in,  but implementation did not start yet. 
Well, contribution to any of above is welcomed.

potatop commented:
Yes, I was talking about the RequirementRule functionality.

qiwzhang commented:
* yes, it is almost the same, but there are different; the former will not do any jwt verification,  the latter will verify all jwts in the request, but allow fails and missing.

* if no requirements matched a path, it will be the same as empty JwtRequirement; not any jwt verification.

lizan commented:
@potatop can you open a new issue for further questions/feature requests?
